### PR TITLE
[SYCLomatic] Fix the usage after dev_mgr::_dev_stack destroyed

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 1]
```diff
@@ -783,20 +783,36 @@ class dev_mgr {
...
+        id = DEFAULT_DEVICE_ID;
+        return;
+      }
       throw std::runtime_error("invalid device id");
     }
   }
+
+  class stack_wrapper : public std::stack<unsigned int> {
+  public:
+    stack_wrapper() { _stack_addr.insert(this); }
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]: add lock

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 2]
```diff
@@ -783,20 +783,36 @@ class dev_mgr {
...
+    ~stack_wrapper() { _stack_addr.erase(this); }
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]: add lock

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 3]
```diff
@@ -783,20 +783,36 @@ class dev_mgr {
...
   std::vector<std::shared_ptr<device_ext>> _devs;
   /// stack of devices resulting from CUDA context change;
-  static inline thread_local std::stack<unsigned int> _dev_stack;
+  static inline thread_local stack_wrapper _dev_stack;
   /// DEFAULT_DEVICE_ID is used, if current_device_id() finds an empty
   /// _dev_stack, which means the default device should be used for the current
   /// thread.
   const unsigned int DEFAULT_DEVICE_ID = 0;
   int _cpu_device = -1;
+  // Add address when constructing _dev_stack, and remove it when destructing
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]:
```suggestion
  // Add address when constructing _dev_stack, and remove it when destructing.
```

**Review by ziranzha — APPROVED** [review]

**Pull request merged**